keps/0018-controller-overhaul.md

```diff
@@ -0,0 +1,121 @@
+
+## Table of Contents
+
+* [Controller redesign](#controller-redesign)
+      * [Table of Contents](#table-of-contents)
+      * [Summary](#summary)
+      * [Motivation](#motivation)
+         * [Goals](#goals)
+         * [Non-Goals](#non-goals)
+      * [Proposal](#proposal)
+         * [Instance CRD changes](#instance-crd-changes)
+         * [Admission webhook](#admission-webhook)
+         * [Implementation details](#implementation-detailsnotesconstraints)
+         * [Risks and mitigations](#risks-and-mitigations)
+      * [Graduation Criteria](#graduation-criteria)
+
+## Summary
+
+The way KUDO controllers are implemented currently has several flaws:
+1. When KUDO manager is down (or restarted) we might miss an update on a CRD meaning we won’t execute the plan that was supposed to run at or we execute incorrect one ([issue](https://github.com/kudobuilder/kudo/issues/422))
+2. Multiple plan can be seemingly in progress leading to misleading status communicated ([issue](https://github.com/kudobuilder/kudo/issues/628))
+3. No way to ensure atomicity on plan execution (given the current CRD design where information is spread across several CRDs)
+4. Very low test coverage and overall confidence in the code
+
+To address all of the above points we’re proposing new design of our CRDs. We’re going to get rid of PlanExecution CRD altogether and merge the execution status into the Instance’s status subresource. At the same time we’re going to keep last known state of instance inside status subresource as well, which will allow us to properly identify plans that are supposed to be run after user submits a change to the Instance spec.
+
+## Motivation
+
+### Goals
+
+- better understandable and tested controller code
+- ensuring correct plan is executed in case KUDO controller is not running for some time
+- atomicity on plan execution level - do not allow to run plan when other plan is in progress
+- be able to see current plan execution status
+
+### Non-Goals
+
+- providing longer history of plan executions that just the last one run
+- have manual plan execution feature (there are several approaches to how to do this, I think it deserves a separate KEP)
+
+## Proposal
+
+The proposed design that will solve all the current known problems with controllers and CRDs is to get rid of PlanExecution CRD altogether. Plan status will basically become instance status (it’s the case right now as well anyway, we just need to replicate those values across two places). 
+
+### Instance CRD changes
+
+The instance CRD for instance with running deploy plan might then look as following.
+
+```yaml
+ Instance
+  spec:
+    operatorVersion: operator-123
+    parameters:
+      - REPLICAS: 4
+  status:
+    planStatus: # (for every plan defined in OV, there will be one field)
+      - deploy:
+          status: IN_PROGRESS
+          executionName: deploy-1478631057
+          lastFinishedRun: 1568642057
+          overview:
+            strategy: serial
```

Comment:
Why is `strategy` on this level (outside `phases`)?

Reply:
because plan has a strategy as well...

Reply:
Of course, `plan.strategy` defines how `phases` are executed 🤦‍♂

Reply:
Next question: why do we need to mention `strategies` in `planStatus` at all? Aren't they part of plan/phase/step definition (like e.g. `resources`) and can be omitted?

Reply:
yeah :smile: I was expecting that question. I think it's just for reporting purposes... same as delete. So e.g. if CLI pulls the status of instance, the strategy is an important thing to show there as well. So we either have it there (which was the case now) or we pull both OV and I in cli and try to merge those two together. I am actually fine with both approaches...

Reply:
I see. Imho, I'd rather have a clear cut between `spec` and `status` and merge them later. (this way, merging is also easy)

Reply:
@zen-dog so you would remove it from status and pull both CRDs in CLI, correct? I am fine with that